jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -2485,15 +2510,15 @@ public boolean regionMatches(boolean ignoreCase, int thisStart, String string, i
 				if (Character.isHighSurrogate(charAtO1) && Character.isHighSurrogate(charAtO2) && (o1 < end)) {
 					int codepointAtO1 = Character.toCodePoint(charAtO1, s1.charAtInternal(o1++, s1Value));
 					int codepointAtO2 = Character.toCodePoint(charAtO2, s2.charAtInternal(o2++, s2Value));
-					if ((codepointAtO1 != codepointAtO2) && (compareValue(codepointAtO1) != compareValue(codepointAtO2))) {
+					if ((codepointAtO1 != codepointAtO2)
+							&& (compareValue(codepointAtO1) != compareValue(codepointAtO2))
+					) {
 						return false;
 					}
 				}
 				/*[ENDIF] JAVA_SPEC_VERSION >= 16 */
 
-				if (charAtO1 != charAtO2 &&
-						toUpperCase(charAtO1) != toUpperCase(charAtO2) &&
-						toLowerCase(charAtO1) != toLowerCase(charAtO2)) {
+				if (!charValuesEqualIgnoreCase(charAtO1, charAtO2)) {
```

Comment:
I think we may be missing the `charAtO1 != charAtO2` check here. I see the check is duplicated in `charValuesEqualIgnoreCase` which seems suboptimal. I think we should take the check out of `charValuesEqualIgnoreCase` and always perform it upfront before calling `charValuesEqualIgnoreCase`, otherwise performance will suck.

Reply:
Moved the `char1 != char2` check back outside of `charValuesEqualIgnoreCase()`